### PR TITLE
Fixup largefile define

### PR DESCRIPTION
**Pull request by AdrianVovk**
Fixes https://github.com/systemd/systemd/issues/31417

Also makes the error messages include the flags that are invalid, for easier debugging in the future, and defines a macro that declares what flags are safe to use (since there's > 1 consumer of this nowadays, and more coming with #26826)

### CODE REVIEW

**Comment by mrc0mmand** [comment]
I tried this on both ppc64le and aarch64 and the issue seems to be indeed gone, thanks!

Also, just out of curiosity, I tried this on armv7l as well, but there the test doesn't get far:

```
[1216026.467820] testsuite-46.sh[61]: + mount -t tmpfs tmpfs /home -o size=290M
[1216026.481155] testsuite-46.sh[61]: + NEWPASSWORD=xEhErW0ndafV4s
[1216026.481155] testsuite-46.sh[61]: + homectl create test-user --disk-size=min --luks-discard=yes --image-path=/home/test-user.home --luks-pbkdf-type=pbkdf2 --luks-pbkdf-time-cost=1ms
[1216026.486666] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/home_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=23 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[1216026.488123] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/home_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=24 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[1216026.489369] systemd[1]: home.mount: Changed dead -> mounted
[1216026.491048] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/home_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=25 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[1216026.492273] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/home_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=26 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[1216026.493456] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/swap_2etarget interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=27 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[1216026.494637] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/umount_2etarget interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=28 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[1216026.495790] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/local_2dfs_2dpre_2etarget interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=29 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[1216026.496956] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/local_2dfs_2etarget interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=30 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[1216027.126558] systemd-homed[53]: Got message type=method_call sender=:1.5 destination=org.freedesktop.home1 path=/org/freedesktop/home1 interface=org.freedesktop.home1.Manager member=CreateHomeEx  cookie=2 reply_cookie=0 signature=sa{sh}t error-name=n/a error-message=n/a
[1216027.128567] systemd-homed[53]: Sent message type=error sender=n/a destination=:1.5 path=n/a interface=n/a member=n/a cookie=5 reply_cookie=2 signature=s error-name=org.freedesktop.DBus.Error.NotSupported error-message=Provided flags are unsupported.
[1216027.130249] systemd-homed[53]: Failed to process message type=method_call sender=:1.5 destination=org.freedesktop.home1 path=/org/freedesktop/home1 interface=org.freedesktop.home1.Manager member=CreateHomeEx cookie=2 reply_cookie=0 signature=sa{sh}t error-name=n/a error-message=n/a: Provided flags are unsupported.
[1216027.131788] homectl[66]: Operation on home test-user failed: Provided flags are unsupported.
```

**Review by bluca — APPROVED** [review]

**Review by yuwata — APPROVED** [review]

**Comment by mrc0mmand** [comment]
> Also, just out of curiosity, I tried this on armv7l as well, but there the test doesn't get far:
> 
> ```
> [1216026.467820] testsuite-46.sh[61]: + mount -t tmpfs tmpfs /home -o size=290M
> [1216026.481155] testsuite-46.sh[61]: + NEWPASSWORD=xEhErW0ndafV4s
> [1216026.481155] testsuite-46.sh[61]: + homectl create test-user --disk-size=min --luks-discard=yes --image-path=/home/test-user.home --luks-pbkdf-type=pbkdf2 --luks-pbkdf-time-cost=1ms
> [1216026.486666] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/home_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=23 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
> [1216026.488123] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/home_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=24 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
> [1216026.489369] systemd[1]: home.mount: Changed dead -> mounted
> [1216026.491048] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/home_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=25 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
> [1216026.492273] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/home_2emount interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=26 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
> [1216026.493456] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/swap_2etarget interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=27 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
> [1216026.494637] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/umount_2etarget interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=28 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
> [1216026.495790] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/local_2dfs_2dpre_2etarget interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=29 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
> [1216026.496956] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/local_2dfs_2etarget interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=30 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
> [1216027.126558] systemd-homed[53]: Got message type=method_call sender=:1.5 destination=org.freedesktop.home1 path=/org/freedesktop/home1 interface=org.freedesktop.home1.Manager member=CreateHomeEx  cookie=2 reply_cookie=0 signature=sa{sh}t error-name=n/a error-message=n/a
> [1216027.128567] systemd-homed[53]: Sent message type=error sender=n/a destination=:1.5 path=n/a interface=n/a member=n/a cookie=5 reply_cookie=2 signature=s error-name=org.freedesktop.DBus.Error.NotSupported error-message=Provided flags are unsupported.
> [1216027.130249] systemd-homed[53]: Failed to process message type=method_call sender=:1.5 destination=org.freedesktop.home1 path=/org/freedesktop/home1 interface=org.freedesktop.home1.Manager member=CreateHomeEx cookie=2 reply_cookie=0 signature=sa{sh}t error-name=n/a error-message=n/a: Provided flags are unsupported.
> [1216027.131788] homectl[66]: Operation on home test-user failed: Provided flags are unsupported.
> ```

-> https://github.com/systemd/systemd/pull/31420

File: src/basic/fd-util.h [hunk 1]
```diff
@@ -21,6 +22,20 @@
 #define EBADF_PAIR { -EBADF, -EBADF }
 #define EBADF_TRIPLET { -EBADF, -EBADF, -EBADF }
 
+/* Flags that are safe to have set on an FD given to a privileged service to operate on.
+ * This ensures that clients can't trick a privileged service into giving access to a file the client
+ * doesn't already have access to (especially via something like O_PATH).
+ *
+ * O_NOFOLLOW: For some reason the kernel will return this flag from fcntl; it doesn't go away immediately
```
Comment (AdrianVovk) [inline src/basic/fd-util.h:29]: Btw, this is new w/ this PR. Just double checking that this was noticed

File: src/home/homed-bus.c [hunk 2]
```diff
@@ -120,8 +119,10 @@ int bus_message_read_blobs(sd_bus_message *m, Hashmap **ret, sd_bus_error *error
 
                 /* Refuse fds w/ unexpected flags set. In particular, we don't want to permit O_PATH FDs, since
                  * those don't actually guarantee that the client has access to the file. */
-                if ((flags & ~(O_ACCMODE|RAW_O_LARGEFILE)) != 0)
-                        return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "FD for %s has unexpected flags set", filename);
+                if (UNSAFE_FD_FLAGS(flags) != 0)
+                        return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS,
+                                                 "FD for %s has unexpected flags set: 0%o",
```
Comment (YHNdnzj) [inline src/home/homed-bus.c:124]: So I think this is probably too verbose to be included in bus error. We should add a helper that does log_debug internally. But I'll merge this as-is and submit a follow-up.

**Pull request merged**